Checks for failing URLs

Morzaram opened the pull request:
## Describe your changes
Job will run weekly and will fail if there's a broken link

## Related issue number/link
Fixes https://github.com/ClimateTown/knowledge-hub/issues/14

Thanks for contributing! 🥳🚀🌳

VeckoTheGecko commented:
Heyo @Morzaram , cheers for the PR. Regarding the codebase in general, I've backtracked on my initial stance in [my other comment](https://github.com/ClimateTown/knowledge-hub/issues/14#issuecomment-1477799885) (sorry about that).

For this codebase, I think it would be good to stick to features in Python/Typescript, for a couple reasons:
- Extensibility: Their popularity means that fellow contributors are more likely to know the language, hence iterate on existing code (rather than having to refactor).
- Code maintainence: Having select languages in the codebase reduces maintainence burden (I severely over-estimated my ability to meaningfully review code written in completely unfamiliar languages 😅).

Keen to know what you think about this/how it fits with you. Would still love to have your contributions here

Morzaram commented:
Sire! No problem! I'll push a rewrite soon tomorrow!

VeckoTheGecko commented:
Awesome! If you're going for a Python approach, [asyncio](https://docs.python.org/3/library/asyncio.html) would be good to look into to handle the sending of requests to the URLs (although that can be included in a later PR if need be).

VeckoTheGecko commented on scripts/check_urls/package.json:
Is there any way we can use the same environment that we use for the build process? That already has all the needed dependencies, and may be easier than setting up two environments.

VeckoTheGecko commented on scripts/check_urls/check_urls.ts:
```diff
@@ -0,0 +1,57 @@
+      console.error("The following links are broken:");
+      errorLinks.forEach((link) => {
+        if (link.status == 301) {
+          console.warn(`${link.url} redirects to -> ${link.redirect}`);
```
`response.statusMessage` already contains this information. I think its easier to store `url`, `statusCode`, and `statusMesssage` and then print these back for all URLs.

Maybe:
```ts
console.log(`Code ${link.statusCode} for ${link.url}: ${link.statusMessage}`);
```

VeckoTheGecko commented on scripts/check_urls/check_urls.ts:
```diff
@@ -0,0 +1,57 @@
+
+  const checkUrl = async (url: UrlObject) => {
+    try {
+      const response = await fetch(new URL(url.url).toString(), {
+        method: "HEAD",
+      });
```
```suggestion
      const response = await fetch(new URL(url.url).toString(), {
        method: "HEAD",
        redirect: "manual",
      });
```
would mean that redirects would give response code `301`, and you don't need to look at `.redirected`.

VeckoTheGecko replied:
Actually, I don't think this suggestion would work anyway in the case of multiple redirects. Anyway, the existing implementation is good

VeckoTheGecko commented on scripts/check_urls/check_urls.ts:
```diff
@@ -0,0 +1,57 @@
+
+  Promise.all(urls.map(checkUrl)).then((links) => {
+    links
+      .filter((l) => l.status == 200)
```
All `2XX` status codes are all ok. It would be good to futureproof this in case we get non 200 passing codes.

VeckoTheGecko commented on scripts/check_urls/check_urls.ts:
```diff
@@ -0,0 +1,57 @@
+        console.log(`${link.url} - OK ${link.status}`);
+      });
+
+    const errorLinks = links.filter((l) => l.status != 200);
```
Hmm. This is interesting. `301` codes are only problematic if they direct to a different domain (e.g. `www.350.org` redirecting to `350.org` doesn't matter), as that indicates that they are switching domains and the old one may become a `404` error sometime. Would you be able to implement a check for this?

Morzaram replied:
@VeckoTheGecko I think the new code should resolve it. I do a check where if the domain is the same then it passes as a live link.

VeckoTheGecko reviewed (changes requested):
Thanks for the PR! Just some changes and queries. Also if you wouldn't mind making the script more modular (avoiding large arrow functions, or nested function definitions).

VeckoTheGecko commented:
Fixes #14

Morzaram commented:
@VeckoTheGecko 
1. I rewrote the script quite a bit, so it's more readable. 
2. added comments so you can understand what is happening in each section
3. moved it to top-level package.json. Added a few dev dependencies but nothing for normal dependencies

Hope you like it!

VeckoTheGecko reviewed (approved):
Love these changes! Thanks for addressing the PR comments 😄 🚀
